Add Documentation for Extension SDK Composables (#20684)
directus/directus: docs/extensions/sdk-composables.md

---
description: Learn about the extension SDK composables and how to utilize them when developing custom extensions.
contributors: Esther Agbaje
---

# Extension SDK Composables

> The Extension Composables serve as a powerful toolkit to help developers create extensions in Directus.

Reviewer: ```suggestion
There are several composables available as part of the Directus Extensions SDK that make working with Directus easier.
```

> Rather than needing to rewrite logic from scratch, extension developers can leverage primitives like `useApi()` or
> `useStores()`, to abstract away complexity when building extensions.

Reviewer: ```suggestion
Rather than needing to rewrite logic from scratch, extension developers can leverage primitives like `useApi()` or
`useStores()`, to handle common complexities when building extensions.
```

Reviewer: "Abstract away" works when spoken. Less in docs.

> ## useApi()

Reviewer: ```suggestion
## `useApi()`
```

Reviewer: Can I ask that you do this with all subheadings?

> The `useApi` composable allows extensions to make requests to endpoints within Directus. It acts as a wrapper around the
> axios library.

Reviewer: ```suggestion
The `useApi` composable allows extensions to make requests to endpoints within Directus. It acts as a wrapper around the
axios library. It automatically authenticates requests made from your app extensions with the current user.
```

Reviewer: Needed an explainer of 'why'

Here’s a sample code of how to implement `useApi`

```ts
<script setup>
import { useApi } from '@directus/extensions-sdk';

const api = useApi();

async function fetchData() {

>       const response = await api.get(ENDPOINT-URL);

Reviewer: ```suggestion
      const response = await api.get(ENDPOINT_URL);
```

      data.value = response.data;

>     }

Reviewer: ```suggestion
};
```

fetchData();
</script>
```

## useStores()

`useStores` serves as the main entry point for accessing specific functionality when building extensions. Within

> `useStores` are sub-stores like the `usePermissionsStore`, `useCollectionsStore` and `useFieldsStore` among others.

Reviewer: > accessing specific functionality when building extensions

This doesn't mean anything - can we revise this? If it's specifically about collections, then say this.

Reviewer: Is "sub-stores" the right term?

```ts
<script setup>
import { useStores } from '@directus/extensions-sdk';

const { useFieldsStore, usePermissionsStore, useCollectionStore } = useStores();
const fieldsStore = useFieldsStore();
const permissionsStore = usePermissionsStore();
const collectionStore = useCollectionStore();
</script>
```

### useFieldsStore

The `useFieldsStore` gives developers access to customizable field properties and functionality. It provides methods
like `fields`, `updateField` , and `getPrimaryKeyFieldForCollection`.

For a deep dive on how to use the `useFieldsStore` composable,

> [see the implementation](https://github.com/directus/directus/blob/main/app/src/stores/fields.ts#L69) in our codebase.

Reviewer: This is missing the 'why'. WHY does this exist? WHEN would I use it? Same for all items.
